### PR TITLE
[IMP] project: add creation of analytic accounts in batch

### PR DESCRIPTION
**Pull request by dise-odoo**
When creating a project, an analytic account is created. However, when creating projects in batch, each analytic account is created separately, when they could be created in batch for optimization.

This PR adds a _get_values_analytic_account_batch() method to help with the creation of analytic accounts in batch.

task-3820424

### CODE REVIEW

File: addons/project/models/project_project.py [hunk 1]
```diff
@@ -857,29 +857,25 @@ def _get_hide_partner(self):
...
     @api.model
-    def _create_analytic_account_from_values(self, values):
-        company = self.env['res.company'].browse(values.get('company_id', False))
+    def _get_values_analytic_account_batch(self, project_vals):
         project_plan_id = int(self.env['ir.config_parameter'].sudo().get_param('analytic.analytic_plan_projects'))
-
         if not project_plan_id:
             project_plan, _other_plans = self.env['account.analytic.plan']._get_all_plans()
             project_plan_id = project_plan.id
+        companies = self.env['res.company'].browse([project.get('company_id', False) for project in project_vals])
```
Comment (damrOdoo) [inline addons/project/models/project_project.py:865]: i'd suggest renaming the 'project' into 'val', or at least 'project_val'. Using 'project' implicitly imply its a project record imo, while its just a dict of vals.

File: addons/project/models/project_project.py [hunk 2]
```diff
@@ -857,29 +857,25 @@ def _get_hide_partner(self):
         return False
 
     @api.model
-    def _create_analytic_account_from_values(self, values):
-        company = self.env['res.company'].browse(values.get('company_id', False))
+    def _get_values_analytic_account_batch(self, project_vals):
         project_plan_id = int(self.env['ir.config_parameter'].sudo().get_param('analytic.analytic_plan_projects'))
-
```
Comment (damrOdoo) [inline addons/project/models/project_project.py:863]: unnecessary diff :angel:

File: addons/project/models/project_project.py [hunk 3]
```diff
@@ -857,29 +857,25 @@ def _get_hide_partner(self):
...
+        } for project, company in zip(project_vals, companies)]
 
     def _create_analytic_account(self):
-        for project in self:
-            project.analytic_account_id = self._create_analytic_account_from_values({
-                'company_id': project.company_id.id,
-                'name': project.name,
-                'partner_id': project.partner_id.id
-            })
+        analytic_accounts_values = self._get_values_analytic_account_batch(self.read(['name', 'company_id', 'partner_id'], None))
```
Comment (damrOdoo) [inline addons/project/models/project_project.py:875]: I never thought of using read that way. It's interesting. :monocle_face:

File: addons/hr_timesheet/models/project_project.py [hunk 4]
```diff
@@ -155,20 +155,20 @@ def create(self, vals_list):
...
+            vals['analytic_account_id'] = analytic_account.id
         return super().create(vals_list)
 
     def write(self, values):
         # create the AA for project still allowing timesheet
         if values.get('allow_timesheets') and not values.get('analytic_account_id'):
-            for project in self:
-                if not project.analytic_account_id:
-                    project._create_analytic_account()
+            create_account = self.filtered(lambda project: not project.analytic_account_id)
```
[the remark this answers is not in the document]
Comment (dise-odoo) [reply addons/hr_timesheet/models/project_project.py:170]: Yep you are right, let's go with this :)

**Review by damrOdoo — COMMENTED** [review]
Thanks for your fix. I left one last comment and it'll be ok.

File: addons/project/tests/test_project_create_batch.py
[the remark this answers is not in the document]
Comment (dise-odoo) [reply addons/project/tests/test_project_create_batch.py:1]: Ah yes, but how do I tag it as post install though?

[the remark this answers is not in the document]
Comment (dise-odoo) [reply addons/project/tests/test_project_create_batch.py:1]: After rethinking, I'd rather not have it in post install :)

**Comment by damrOdoo** [comment]
lgtm. Thanks for your work.

File: addons/project/models/project_project.py [hunk 5]
```diff
@@ -857,29 +857,26 @@ def _get_hide_partner(self):
...
+        } for val, company in zip(project_vals, companies)]
 
     def _create_analytic_account(self):
-        for project in self:
-            project.analytic_account_id = self._create_analytic_account_from_values({
-                'company_id': project.company_id.id,
-                'name': project.name,
-                'partner_id': project.partner_id.id
-            })
+        analytic_accounts_values = self._get_values_analytic_account_batch(self.read(['name', 'company_id', 'partner_id'], None))
```
Comment (xavierbol) [inline addons/project/models/project_project.py:876]: We could use `_read_format` to avoid fetching the data with the read if we already prefetch the data. 🙂 
```suggestion
        analytic_accounts_values = self._get_values_analytic_account_batch(self._read_format(['name', 'company_id', 'partner_id'], None))
```

FYI @damrOdoo 😁

File: addons/project/models/project_project.py [hunk 6]
```diff
@@ -857,29 +857,26 @@ def _get_hide_partner(self):
...
+        analytic_accounts = self.env['account.analytic.account'].create(analytic_accounts_values)
+        for project, analytic_account in zip(self, analytic_accounts):
+            project.analytic_account_id = analytic_account.id
```
Comment (xavierbol) [inline addons/project/models/project_project.py:879]: Equivalent to:
```suggestion
            project.analytic_account_id = analytic_account
```
🙂 

In the write method indeed we need to give the id but just for an update we can give the recordset. 🙂

**Review by xavierbol — COMMENTED** [review]
LGTM, thanks for your work, I left 2 comments but I already applied them since I checkout your branch to do some additional tests. 😁

**Comment by xavierbol** [comment]
robodoo r+
